# Ridge Regression
###### credit to dataquestio and VikParuchuri

##### importing pandas library

In [2]:
import pandas as pd

###### using the pandas read csv function to read in nba_games.csv
###### our first column is pandas index because index_col=0

In [3]:
dataFrame = pd.read_csv("nba_games.csv", index_col=0)

###### this cahnges the index to another set of values 
###### drop = True makes sure that the old index is not added to the columns
###### the new index is kept but the old one is replaced
###### when working with timeseries data the index should either follow the same order as the date or be the date 

In [4]:
dataFrame = dataFrame.reset_index(drop = True)

###### sort data frame in order of date which makes processing easier later on

In [5]:
dataFrame = dataFrame.sort_values("date")

In [6]:
dataFrame = dataFrame.reset_index(drop = True)

###### the bold values are the index value which are assigned to each row
###### this is what lets us call the row by its index
###### each row is one nba game each column is the stats for that game
###### the end columns are meta data about seasons date and who won

In [7]:
dataFrame 

mp  mp.1  fg  fga    fg%  3p  3pa    3p%  ft  fta  ...  tov%_max_opp  \
0      240   240  33   87  0.379   5   16  0.313  22   29  ...          36.8   
1      240   240  43   92  0.467   8   26  0.308   9   10  ...          33.3   
2      240   240  36   91  0.396  15   33  0.455  17   18  ...          31.7   
3      240   240  31   84  0.369   6   19  0.316  13   17  ...          50.0   
4      240   240  34   85  0.400  10   32  0.313  16   18  ...          43.6   
...    ...   ...  ..  ...    ...  ..  ...    ...  ..  ...  ...           ...   
17767  240   240  43   97  0.443  16   38  0.421  19   21  ...          16.8   
17768  240   240  47   85  0.553  19   41  0.463  26   30  ...          26.2   
17769  240   240  36  102  0.353  12   42  0.286  22   27  ...          24.0   
17770  240   240  41  100  0.410  11   35  0.314  15   16  ...          37.5   
17771  240   240  35   86  0.407  12   27  0.444  26   30  ...          50.0   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  \
0              40.4           150           105       PHI         84   
1              44.9           200           120       ORL         91   
2              30.6           174           130       CHO         94   
3              42.7           195           102       CHI        108   
4              25.8           150           120       TOR        104   
...             ...           ...           ...       ...        ...   
17767          36.4           160           137       CLE        118   
17768          31.1           165           146       POR        106   
17769          46.7           286           117       LAL        139   
17770          34.3           178           117       BOS        123   
17771          50.2           217           120       IND        106   

       home_opp  season        date    won  
0             0    2016  01/01/2016   True  
1             0    2016  01/01/2016   True  
2             0    2016  01/01/2016   True  
3             1    2016  01/01/2016  False  
4             1    2016  01/01/2016  False  
...         ...     ...         ...    ...  
17767         1    2022  31/12/2021   True  
17768         0    2022  31/12/2021   True  
17769         1    2022  31/12/2021  False  
17770         1    2022  31/12/2021  False  
17771         1    2022  31/12/2021   True  

[17772 rows x 150 columns]

###### deleting the columns that are duplicates and columns that i dont need

In [8]:
del dataFrame["mp.1"]
del dataFrame["mp_opp.1"]
del dataFrame["index_opp"]

In [9]:
dataFrame

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  tov%_max_opp  \
0      240  33   87  0.379   5   16  0.313  22   29  0.759  ...          36.8   
1      240  43   92  0.467   8   26  0.308   9   10  0.900  ...          33.3   
2      240  36   91  0.396  15   33  0.455  17   18  0.944  ...          31.7   
3      240  31   84  0.369   6   19  0.316  13   17  0.765  ...          50.0   
4      240  34   85  0.400  10   32  0.313  16   18  0.889  ...          43.6   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  43   97  0.443  16   38  0.421  19   21  0.905  ...          16.8   
17768  240  47   85  0.553  19   41  0.463  26   30  0.867  ...          26.2   
17769  240  36  102  0.353  12   42  0.286  22   27  0.815  ...          24.0   
17770  240  41  100  0.410  11   35  0.314  15   16  0.938  ...          37.5   
17771  240  35   86  0.407  12   27  0.444  26   30  0.867  ...          50.0   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  \
0              40.4           150           105       PHI         84   
1              44.9           200           120       ORL         91   
2              30.6           174           130       CHO         94   
3              42.7           195           102       CHI        108   
4              25.8           150           120       TOR        104   
...             ...           ...           ...       ...        ...   
17767          36.4           160           137       CLE        118   
17768          31.1           165           146       POR        106   
17769          46.7           286           117       LAL        139   
17770          34.3           178           117       BOS        123   
17771          50.2           217           120       IND        106   

       home_opp  season        date    won  
0             0    2016  01/01/2016   True  
1             0    2016  01/01/2016   True  
2             0    2016  01/01/2016   True  
3             1    2016  01/01/2016  False  
4             1    2016  01/01/2016  False  
...         ...     ...         ...    ...  
17767         1    2022  31/12/2021   True  
17768         0    2022  31/12/2021   True  
17769         1    2022  31/12/2021  False  
17770         1    2022  31/12/2021  False  
17771         1    2022  31/12/2021   True  

[17772 rows x 147 columns]

In [10]:
#this pulls the "won" column from the next game and pull it back one row
def get_target(team):
    team["target"] = team["won"].shift(-1)
    return team

#this splits up the data frame by teams before the next row is pulled back to become the target
#this gets the specific teams next game
dataFrame = dataFrame.groupby("team", group_keys=False).apply(get_target)

###### this takes only the rows from the data frame where the team code is GSW
###### shows all the rows for the Golden State Warriors where each row is the games they played
###### the target indicates if they won or lost their next game
###### the won column indicates if they have won or lost there current game

In [11]:
dataFrame[dataFrame["team"] == "GSW"]

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  usg%_max_opp  \
54     240  31   79  0.392   7   35  0.200  29   33  0.879  ...          29.0   
76     240  46   86  0.535  17   32  0.531  14   21  0.667  ...          29.2   
97     240  47   96  0.490  21   48  0.438  11   14  0.786  ...          30.8   
147    240  48   90  0.533  18   35  0.514  17   19  0.895  ...          28.7   
187    240  43   96  0.448  20   46  0.435  18   18  1.000  ...          34.6   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17621  240  46   84  0.548  17   30  0.567  25   35  0.714  ...          36.1   
17642  240  47   89  0.528  16   32  0.500  21   30  0.700  ...          26.8   
17679  240  43   84  0.512  11   23  0.478  17   27  0.630  ...          34.3   
17712  240  46   84  0.548  14   28  0.500  26   32  0.813  ...          44.5   
17727  265  44   96  0.458  12   27  0.444  13   18  0.722  ...          25.9   

       ortg_max_opp  drtg_max_opp  team_opp  total_opp  home_opp  season  \
54              250           102       POR        123         0    2021   
76              163           127       UTA        116         1    2022   
97              231           131       CHO        111         0    2017   
147             139           139       CLE        112         1    2020   
187             230           138       SAS        120         1    2022   
...             ...           ...       ...        ...       ...     ...   
17621           218           131       NOP        120         1    2016   
17642           149           132       NOP        121         0    2019   
17679           157           124       HOU        110         1    2016   
17712           137           134       PHO        109         1    2019   
17727           175           120       SAS        117         1    2020   

             date    won  target  
54     01/01/2021  False    True  
76     01/01/2022   True    True  
97     01/02/2017   True    True  
147    01/02/2020   True    True  
187    01/02/2022   True    True  
...           ...    ...     ...  
17621  31/10/2015   True    True  
17642  31/10/2018   True    True  
17679  31/12/2015   True    True  
17712  31/12/2018   True   False  
17727  31/12/2019  False     NaN  

[655 rows x 148 columns]

###### will show which teams played in the game with a 3 letter abreviation

In [2]:
dataFrame["team"]

NameError: name 'dataFrame' is not defined

###### finds all of the null values in the target column 
###### the pandas isnull function will return true if a value is null false if it isn't
###### this gives all the null values in the target column
###### for example when there is no next game
###### replace all the missing rows with the number 2

In [13]:
dataFrame["target"][pd.isnull(dataFrame["target"])] = 2

C:\Users\Garnet\AppData\Local\Temp\ipykernel_20912\991180924.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataFrame["target"][pd.isnull(dataFrame["target"])] = 2


###### this is going to change a false in the target column into 0 and a true into 1
###### if there is an error converting the boolean into an integer errors="ignore" igonres this

In [14]:
dataFrame["target"] = dataFrame["target"].astype(int, errors="ignore")

###### 0 means a loos
###### 1 means a win 
###### 2 means there is no data for the next game because this is the last game we have data for for that team

In [15]:
dataFrame

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  usg%_max_opp  \
0      240  33   87  0.379   5   16  0.313  22   29  0.759  ...          40.4   
1      240  43   92  0.467   8   26  0.308   9   10  0.900  ...          44.9   
2      240  36   91  0.396  15   33  0.455  17   18  0.944  ...          30.6   
3      240  31   84  0.369   6   19  0.316  13   17  0.765  ...          42.7   
4      240  34   85  0.400  10   32  0.313  16   18  0.889  ...          25.8   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  43   97  0.443  16   38  0.421  19   21  0.905  ...          36.4   
17768  240  47   85  0.553  19   41  0.463  26   30  0.867  ...          31.1   
17769  240  36  102  0.353  12   42  0.286  22   27  0.815  ...          46.7   
17770  240  41  100  0.410  11   35  0.314  15   16  0.938  ...          34.3   
17771  240  35   86  0.407  12   27  0.444  26   30  0.867  ...          50.2   

       ortg_max_opp  drtg_max_opp  team_opp  total_opp  home_opp  season  \
0               150           105       PHI         84         0    2016   
1               200           120       ORL         91         0    2016   
2               174           130       CHO         94         0    2016   
3               195           102       CHI        108         1    2016   
4               150           120       TOR        104         1    2016   
...             ...           ...       ...        ...       ...     ...   
17767           160           137       CLE        118         1    2022   
17768           165           146       POR        106         0    2022   
17769           286           117       LAL        139         1    2022   
17770           178           117       BOS        123         1    2022   
17771           217           120       IND        106         1    2022   

             date    won  target  
0      01/01/2016   True       0  
1      01/01/2016   True       0  
2      01/01/2016   True       1  
3      01/01/2016  False       0  
4      01/01/2016  False       0  
...           ...    ...     ...  
17767  31/12/2021   True       2  
17768  31/12/2021   True       2  
17769  31/12/2021  False       2  
17770  31/12/2021  False       2  
17771  31/12/2021   True       2  

[17772 rows x 148 columns]

In [16]:
#this is to see if the "won" column is balanced
#there is the same amount of the team lost as the teams won
#this is because NBA games dont have ties, you either win or lose
dataFrame["won"].value_counts()

True     8886
False    8886
Name: won, dtype: int64

In [17]:
dataFrame["target"].value_counts()
#2 values make the target column not balanced
#this is because sometimes the teams do not have a next game 
#this is coded as a 2 instead of a win(1) or a loss(0)

1    8872
0    8870
2      30
Name: target, dtype: int64

In [18]:
#removing null values for the machine learning model to work better
#checking how many columns have null values 
#the pandas isnull function checks if each entry in the data frame is null or not
nulls = pd.isnull(dataFrame)

In [19]:
#the sum function finds the count of the trues in each column
#this is therefore the count of the null values in each column
nulls = nulls.sum()

In [20]:
#this looks at the columns where the count of the nulls is greater than 0
nulls = nulls[nulls > 0]

In [21]:
#looks at nulls
#each cell has false if it is not null 
#and true if it is null
#... in the middle is because pandas truncated or shortened the output and we dont see all the rows
#there are only 6 columns that appear to be all null values
#these columns will be deleted 
nulls

+/-             17772
mp_max          17772
mp_max.1        17772
+/-_opp         17772
mp_max_opp      17772
mp_max_opp.1    17772
dtype: int64

In [22]:
#create a list of valid columns that dont have any null values in them
#these colums are selected from the data frame
#then indexing the pandas index by selecting columns that are not in nulls.index
# ~ is a negation operator or pandas Series
#checking to see if the name of the data frame columns are one of the null columns above
#the column is only kept if it isn't and getting rid of the null columnms and assign this to valid_columns
valid_columns = dataFrame.columns[~dataFrame.columns.isin(nulls.index)]

In [23]:
valid_columns
#this is a pandas index that works like a list

Index(['mp', 'fg', 'fga', 'fg%', '3p', '3pa', '3p%', 'ft', 'fta', 'ft%',
       ...
       'usg%_max_opp', 'ortg_max_opp', 'drtg_max_opp', 'team_opp', 'total_opp',
       'home_opp', 'season', 'date', 'won', 'target'],
      dtype='object', length=142)

In [24]:
dataFrame = dataFrame[valid_columns].copy()
#only select the valid columns for our data frame
#if you assign slices of a data frame back to itself you can get warnings down the line
#.copy avoids this by creating a new data frame and not slices of an exsisting one

In [25]:
dataFrame
#less columns
#data cleaning

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  usg%_max_opp  \
0      240  33   87  0.379   5   16  0.313  22   29  0.759  ...          40.4   
1      240  43   92  0.467   8   26  0.308   9   10  0.900  ...          44.9   
2      240  36   91  0.396  15   33  0.455  17   18  0.944  ...          30.6   
3      240  31   84  0.369   6   19  0.316  13   17  0.765  ...          42.7   
4      240  34   85  0.400  10   32  0.313  16   18  0.889  ...          25.8   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  43   97  0.443  16   38  0.421  19   21  0.905  ...          36.4   
17768  240  47   85  0.553  19   41  0.463  26   30  0.867  ...          31.1   
17769  240  36  102  0.353  12   42  0.286  22   27  0.815  ...          46.7   
17770  240  41  100  0.410  11   35  0.314  15   16  0.938  ...          34.3   
17771  240  35   86  0.407  12   27  0.444  26   30  0.867  ...          50.2   

       ortg_max_opp  drtg_max_opp  team_opp  total_opp  home_opp  season  \
0               150           105       PHI         84         0    2016   
1               200           120       ORL         91         0    2016   
2               174           130       CHO         94         0    2016   
3               195           102       CHI        108         1    2016   
4               150           120       TOR        104         1    2016   
...             ...           ...       ...        ...       ...     ...   
17767           160           137       CLE        118         1    2022   
17768           165           146       POR        106         0    2022   
17769           286           117       LAL        139         1    2022   
17770           178           117       BOS        123         1    2022   
17771           217           120       IND        106         1    2022   

             date    won  target  
0      01/01/2016   True       0  
1      01/01/2016   True       0  
2      01/01/2016   True       1  
3      01/01/2016  False       0  
4      01/01/2016  False       0  
...           ...    ...     ...  
17767  31/12/2021   True       2  
17768  31/12/2021   True       2  
17769  31/12/2021  False       2  
17770  31/12/2021  False       2  
17771  31/12/2021   True       2  

[17772 rows x 142 columns]

In [26]:
#scikit-learn library for machine learning 
#high number of columns can cause problems in machine learning
#there can be columns that are correlated
#some of the columns can cause overfitting or other problems with the model 
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [27]:
#class to split data up properly for feature selector
#cross validation is a strategy for creating predictions across the data set without making predictions on the same data the algorithm is trained on 
#this ensures past data is used to predict the future
from sklearn.model_selection import TimeSeriesSplit
#this feature selector is used to select a smaller set of columns to train the model
from sklearn.feature_selection import SequentialFeatureSelector
#this uses ridge regression to do classification and classify if the team is going to win or lose their next game
from sklearn.linear_model import RidgeClassifier

#initialise ridge classifier
#ridge regression works best when we scale the data subtracting the mean and dividing by standard deviation 
#from further scaling all the values will fall between 0 and 1 
#ratios becaome easier now than if the mean was 0
rc = RidgeClassifier(alpha=1)
#initialise time series split
tss = TimeSeriesSplit(n_splits=3)

#initialise sequential sequence selector
#a feature is a column so 142 features/columns
#the selector is going to train the model using sets of features
#the selector after picking a feature will then test how well the model does and so on
#picked 30 features to select which can be changegd here
#forward in the selector will start from 0 then pick the feature that improves the model the most then keep on doing that till we get to 30
#finally time series split is then used for cross validation
sfs = SequentialFeatureSelector(rc, n_features_to_select=30, direction = "forward", cv=tss)

In [28]:
#these are the columns that are not going to be scaled
removed_columns = ["season","date","won","target","team","team_opp"]

In [29]:
#this selects all of the columns except the columns that are in the removed columns list
#every column is picked in this data frame except the removed_columns
selected_columns = dataFrame.columns[~dataFrame.columns.isin(removed_columns)]

In [30]:
#import MiniMaxScalar
from sklearn.preprocessing import MinMaxScaler

#define scalar
scaler = MinMaxScaler()
#scaling selected columns 
#picking only the selected columns from the data frame
#.fit_transform is a method on the scralar and pass in data fame selected columns
dataFrame[selected_columns] = scaler.fit_transform(dataFrame[selected_columns])

In [31]:
dataFrame
#columns such as field goals (fg) or minuites played (mp)
#the minimum value being 0 and the maximum value being 1 makes ridge regression preform better
#columns that were not supposed to be scaled like date have stayed the same

mp        fg       fga       fg%        3p       3pa       3p%  \
0      0.0  0.318182  0.397059  0.263158  0.172414  0.181818  0.371734   
1      0.0  0.545455  0.470588  0.473684  0.275862  0.333333  0.365796   
2      0.0  0.386364  0.455882  0.303828  0.517241  0.439394  0.540380   
3      0.0  0.272727  0.352941  0.239234  0.206897  0.227273  0.375297   
4      0.0  0.340909  0.367647  0.313397  0.344828  0.424242  0.371734   
...    ...       ...       ...       ...       ...       ...       ...   
17767  0.0  0.545455  0.544118  0.416268  0.551724  0.515152  0.500000   
17768  0.0  0.636364  0.367647  0.679426  0.655172  0.560606  0.549881   
17769  0.0  0.386364  0.617647  0.200957  0.413793  0.575758  0.339667   
17770  0.0  0.500000  0.588235  0.337321  0.379310  0.469697  0.372922   
17771  0.0  0.363636  0.382353  0.330144  0.413793  0.348485  0.527316   

             ft       fta       ft%  ...  usg%_max_opp  ortg_max_opp  \
0      0.488372  0.444444  0.718786  ...      0.234917      0.289100   
1      0.186047  0.142857  0.883314  ...      0.292683      0.526066   
2      0.372093  0.269841  0.934656  ...      0.109114      0.402844   
3      0.279070  0.253968  0.725788  ...      0.264442      0.502370   
4      0.348837  0.269841  0.870478  ...      0.047497      0.289100   
...         ...       ...       ...  ...           ...           ...   
17767  0.418605  0.317460  0.889148  ...      0.183569      0.336493   
17768  0.581395  0.460317  0.844807  ...      0.115533      0.360190   
17769  0.488372  0.412698  0.784131  ...      0.315789      0.933649   
17770  0.325581  0.238095  0.927655  ...      0.156611      0.421801   
17771  0.581395  0.460317  0.844807  ...      0.360719      0.606635   

       drtg_max_opp  team_opp  total_opp  home_opp  season        date    won  \
0          0.329412       PHI   0.192308       0.0    2016  01/01/2016   True   
1          0.505882       ORL   0.259615       0.0    2016  01/01/2016   True   
2          0.623529       CHO   0.288462       0.0    2016  01/01/2016   True   
3          0.294118       CHI   0.423077       1.0    2016  01/01/2016  False   
4          0.505882       TOR   0.384615       1.0    2016  01/01/2016  False   
...             ...       ...        ...       ...     ...         ...    ...   
17767      0.705882       CLE   0.519231       1.0    2022  31/12/2021   True   
17768      0.811765       POR   0.403846       0.0    2022  31/12/2021   True   
17769      0.470588       LAL   0.721154       1.0    2022  31/12/2021  False   
17770      0.470588       BOS   0.567308       1.0    2022  31/12/2021  False   
17771      0.505882       IND   0.403846       1.0    2022  31/12/2021   True   

       target  
0           0  
1           0  
2           1  
3           0  
4           0  
...       ...  
17767       2  
17768       2  
17769       2  
17770       2  
17771       2  

[17772 rows x 142 columns]

In [32]:
#the fit method is used on our selector selected columns are then passed in and the target is passed in 
#this selects the 30 best features to select a target 
sfs.fit(dataFrame[selected_columns], dataFrame["target"])
#representation of the selector

SequentialFeatureSelector(cv=TimeSeriesSplit(gap=0, max_train_size=None, n_splits=3, test_size=None),
                          estimator=RidgeClassifier(alpha=1),
                          n_features_to_select=30)

In [33]:
#get the list of the predictor columns out of the sequential feature selector 
#take the selected columns list into our selector then call the sfs.get_support()
#if the feature selector has picked a feature and thinks this feature should be used in the model sfs.get_support() will be true if not this method will be false
#indexing the selected columns by this will give the list of columns that should be used 
predictors = list(selected_columns[sfs.get_support()])

In [34]:
predictors
#the columns that should be used for predictions
#game stats such as team shooting stats
#individual stats such as the player on the team who got the highest free throws percentage
#opponent team stats and opponent player stats 
#continue at 25:28

['mp',
 'ft',
 'blk',
 'ts%',
 'blk%',
 'usg%',
 'ortg',
 'drtg',
 'ft%_max',
 'tov_max',
 'ts%_max',
 'stl%_max',
 'mp_opp',
 'fg_opp',
 'fga_opp',
 'fg%_opp',
 '3pa_opp',
 'ft_opp',
 'pf_opp',
 '3par_opp',
 'usg%_opp',
 'ortg_opp',
 'drtg_opp',
 'fg_max_opp',
 'fta_max_opp',
 'drb_max_opp',
 'pf_max_opp',
 'pts_max_opp',
 '3par_max_opp',
 'ortg_max_opp']

The function "backtest" does the work of making the predictions because i am using time series data I need to make sure that i am using historical data to predict future data. This function splits the data up by seasons and uses the past seasons to predict the future seasons. As the start parameter is set to 2 this function will need at least 2 seasons of data to start predicting future seasons. For example 2016-17 data would be used to predict 2018 data. Then 2016-18 data will be used to predict 2019 data and so on

If you wanted to make predictions for two seasons at a time you would set the step parameter to 2.


I then fit my model which takes the training data so all of my predictors which is what i am actually giving the model to make its judgement. Then there is the target which is the actual thing that I am trying to predict this is fed into the model so it can figure out which predictors correlate to the target.

Making predictions on the same data I train on will make the model look really good but its unrealistic.

In [35]:
sorted(dataFrame["season"].unique())

[2016, 2017, 2018, 2019, 2020, 2021, 2022]

In [36]:
def backtest(data, model, predictors, start=2, step=1):
    #a list of data frames where each data frame is a 
    #prediction for a single season.
    all_predictions = []
    
    #seasons will be a list of all the seasons that exist in the data
    seasons = sorted(data["season"].unique())
    
    #used to loop through the seasons
    #start parameters default value of 2 can be overrun when you call the backtest function
    #len will be the length of the seasons 
    #the step parameter is used to progress through seasons
    for i in range(start, len(seasons), step):
        #pulls out the season number
        season = seasons[i]
        
        #splits up the training data from the test data
        #training data is used to train my machine learning algorithm which 
        #will use all of the data that comes before the current season 
        train =data[data["season"] < season]
        #the test data is what I will use to generate predictions which 
        #is going to be on the current season
        test = data[data["season"] == season]
        
        #fit model with training data and predictions which is what 
        #the model will use to make its judgements
        #the target is the actual thing I am trying to predict
        #the target is used to figuure out which predictors actually correalte to the target
        model.fit(train[predictors], train["target"])
        
        #generate my predictions by making predictions on the test set
        #by default returns a numPy array
        preds = model.predict(test[predictors])
        #converts numPy array into a Pandas Series
        #I keep the index the same as the test set index
        #this makes it easier to combine the actuals and predictions
        preds = pd.Series(preds, index=test.index)
        
        #creates a Data Frame called combined
        #the pandas concatonate function which puts two series or data frames together
        #the actual values the target is put together with the predictions
        #axis=1 will treat target and predictions as seperate columns thus making 2 columns
        combined = pd.concat([test["target"], preds], axis=1)
        #renaming the two columns as "actual" and "prediction"
        combined.columns = ["actual", "prediction"]
        
        #append the combined data frame to the list all_predictions 
        #all predictions is going to be a list of 5 different Data Frames
        all_predictions.append(combined)
        #combines all 5 data frames into one 
        #pass in the axis parameter 
        #all these data frames have the same column but we want to combine them as seperate rows 
    return pd.concat(all_predictions)

Machine Learning Below

In [37]:
#call the backtest function and pass in data and ridge classifier model and predictors 
predictions = backtest(dataFrame, rc, predictors)

In [38]:
predictions

actual  prediction
20          0           1
21          1           0
22          1           1
23          1           1
24          0           1
...       ...         ...
17767       2           1
17768       2           1
17769       2           0
17770       2           0
17771       2           1

[12522 rows x 2 columns]

In [39]:
from sklearn.metrics import accuracy_score

#this removes the 2s which are the games with no following data 
predictions = predictions[predictions["actual"]!=2]
#this gives the percentage of the time that our prediction was correct 
accuracy_score(predictions["actual"], predictions["prediction"])

0.5052825356170962

I can predict correctly 50.5% of the time which NBA team is going to win a Game 

In [40]:
#the home team is much more likely to win 
#using how often the team is at home as the base line accuracy 
#what percentage of the time the team wins based on whether they are at home or not
#select all the rows where the team won, the count of this and then divide by the total number of rows 
dataFrame.groupby("home").apply(lambda x: x[x["won"] == 1].shape[0]/x.shape[0])
#this splits the data frame into 2 groups one where the team will be away and home 
#then this function will find the winning percentage of both of these groups 

home
0.0    0.428314
1.0    0.571686
dtype: float64

The winning percentage of a team is 57% when the team is at home 
by predicting that the home team will win we will get good accuracy.
This is the bench mark to beat at 57%  

Rolling Averages - The team could have had a random game where the team plays differently for whatever reason. Taking an average of how the team performed in the last x amount of games and use that to give my algorithm information it can use to predict how the team will perform in their next game 

In [41]:
dataFrame

mp        fg       fga       fg%        3p       3pa       3p%  \
0      0.0  0.318182  0.397059  0.263158  0.172414  0.181818  0.371734   
1      0.0  0.545455  0.470588  0.473684  0.275862  0.333333  0.365796   
2      0.0  0.386364  0.455882  0.303828  0.517241  0.439394  0.540380   
3      0.0  0.272727  0.352941  0.239234  0.206897  0.227273  0.375297   
4      0.0  0.340909  0.367647  0.313397  0.344828  0.424242  0.371734   
...    ...       ...       ...       ...       ...       ...       ...   
17767  0.0  0.545455  0.544118  0.416268  0.551724  0.515152  0.500000   
17768  0.0  0.636364  0.367647  0.679426  0.655172  0.560606  0.549881   
17769  0.0  0.386364  0.617647  0.200957  0.413793  0.575758  0.339667   
17770  0.0  0.500000  0.588235  0.337321  0.379310  0.469697  0.372922   
17771  0.0  0.363636  0.382353  0.330144  0.413793  0.348485  0.527316   

             ft       fta       ft%  ...  usg%_max_opp  ortg_max_opp  \
0      0.488372  0.444444  0.718786  ...      0.234917      0.289100   
1      0.186047  0.142857  0.883314  ...      0.292683      0.526066   
2      0.372093  0.269841  0.934656  ...      0.109114      0.402844   
3      0.279070  0.253968  0.725788  ...      0.264442      0.502370   
4      0.348837  0.269841  0.870478  ...      0.047497      0.289100   
...         ...       ...       ...  ...           ...           ...   
17767  0.418605  0.317460  0.889148  ...      0.183569      0.336493   
17768  0.581395  0.460317  0.844807  ...      0.115533      0.360190   
17769  0.488372  0.412698  0.784131  ...      0.315789      0.933649   
17770  0.325581  0.238095  0.927655  ...      0.156611      0.421801   
17771  0.581395  0.460317  0.844807  ...      0.360719      0.606635   

       drtg_max_opp  team_opp  total_opp  home_opp  season        date    won  \
0          0.329412       PHI   0.192308       0.0    2016  01/01/2016   True   
1          0.505882       ORL   0.259615       0.0    2016  01/01/2016   True   
2          0.623529       CHO   0.288462       0.0    2016  01/01/2016   True   
3          0.294118       CHI   0.423077       1.0    2016  01/01/2016  False   
4          0.505882       TOR   0.384615       1.0    2016  01/01/2016  False   
...             ...       ...        ...       ...     ...         ...    ...   
17767      0.705882       CLE   0.519231       1.0    2022  31/12/2021   True   
17768      0.811765       POR   0.403846       0.0    2022  31/12/2021   True   
17769      0.470588       LAL   0.721154       1.0    2022  31/12/2021  False   
17770      0.470588       BOS   0.567308       1.0    2022  31/12/2021  False   
17771      0.505882       IND   0.403846       1.0    2022  31/12/2021   True   

       target  
0           0  
1           0  
2           1  
3           0  
4           0  
...       ...  
17767       2  
17768       2  
17769       2  
17770       2  
17771       2  

[17772 rows x 142 columns]

In [42]:
#data frame called df-rolling
#selected_columns is added into the data frame
#the extra columns "won", "team" and "season" are added to compute the rolling averages
df_rolling = dataFrame[list(selected_columns) + ["won","team","season"]]

In [43]:
df_rolling

mp        fg       fga       fg%        3p       3pa       3p%  \
0      0.0  0.318182  0.397059  0.263158  0.172414  0.181818  0.371734   
1      0.0  0.545455  0.470588  0.473684  0.275862  0.333333  0.365796   
2      0.0  0.386364  0.455882  0.303828  0.517241  0.439394  0.540380   
3      0.0  0.272727  0.352941  0.239234  0.206897  0.227273  0.375297   
4      0.0  0.340909  0.367647  0.313397  0.344828  0.424242  0.371734   
...    ...       ...       ...       ...       ...       ...       ...   
17767  0.0  0.545455  0.544118  0.416268  0.551724  0.515152  0.500000   
17768  0.0  0.636364  0.367647  0.679426  0.655172  0.560606  0.549881   
17769  0.0  0.386364  0.617647  0.200957  0.413793  0.575758  0.339667   
17770  0.0  0.500000  0.588235  0.337321  0.379310  0.469697  0.372922   
17771  0.0  0.363636  0.382353  0.330144  0.413793  0.348485  0.527316   

             ft       fta       ft%  ...  blk%_max_opp  tov%_max_opp  \
0      0.488372  0.444444  0.718786  ...         0.084      0.337526   
1      0.186047  0.142857  0.883314  ...         0.096      0.300839   
2      0.372093  0.269841  0.934656  ...         0.083      0.284067   
3      0.279070  0.253968  0.725788  ...         0.058      0.475891   
4      0.348837  0.269841  0.870478  ...         0.151      0.408805   
...         ...       ...       ...  ...           ...           ...   
17767  0.418605  0.317460  0.889148  ...         0.068      0.127883   
17768  0.581395  0.460317  0.844807  ...         0.045      0.226415   
17769  0.488372  0.412698  0.784131  ...         0.145      0.203354   
17770  0.325581  0.238095  0.927655  ...         0.104      0.344864   
17771  0.581395  0.460317  0.844807  ...         0.051      0.475891   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  total_opp  home_opp    won  \
0          0.234917      0.289100      0.329412   0.192308       0.0   True   
1          0.292683      0.526066      0.505882   0.259615       0.0   True   
2          0.109114      0.402844      0.623529   0.288462       0.0   True   
3          0.264442      0.502370      0.294118   0.423077       1.0  False   
4          0.047497      0.289100      0.505882   0.384615       1.0  False   
...             ...           ...           ...        ...       ...    ...   
17767      0.183569      0.336493      0.705882   0.519231       1.0   True   
17768      0.115533      0.360190      0.811765   0.403846       0.0   True   
17769      0.315789      0.933649      0.470588   0.721154       1.0  False   
17770      0.156611      0.421801      0.470588   0.567308       1.0  False   
17771      0.360719      0.606635      0.505882   0.403846       1.0   True   

       team  season  
0       LAL    2016  
1       WAS    2016  
2       TOR    2016  
3       NYK    2016  
4       CHO    2016  
...     ...     ...  
17767   ATL    2022  
17768   LAL    2022  
17769   POR    2022  
17770   PHO    2022  
17771   CHI    2022  

[17772 rows x 139 columns]

In [44]:
#function to find the team averages 
def find_team_averages(team):
    #data frame is split up into teams
    #then find the performance of that team for the last x amount of games
    #.rolling and .mean pandas function is used to take the last x rows and get a meanof all the columns
    #you can change this around if you want to make predictions earlier in the season like for TRADES
    rolling = team.rolling(10).mean()
    #takes the current game and previous x-1 games 
    #finds the average performance of the team in the previous x amount of games
    return rolling

#this groups by team and season
#the average performance of the previous games is specific to each team
#grouping also by season because teams change a lot from season to season 
#only the respective season should be used
#roup_keys=False stops pandas from adding another level to the index
df_rolling = df_rolling.groupby(["team", "season"], group_keys=False).apply(find_team_averages)#applying find_team_averages function

C:\Users\Garnet\AppData\Local\Temp\ipykernel_20912\1022566045.py:7: FutureWarning: Dropping of nuisance columns in rolling operations is deprecated; in a future version this will raise TypeError. Select only valid columns before calling the operation. Dropped columns were Index(['team'], dtype='object')
  rolling = team.rolling(10).mean()


In [45]:
df_rolling
#NaN is returned from pandas because there are missing values 
#the missing values are because there aren't 10 previous games 
#this can be changed by the number of previous games that are looked at in find_team_averages

mp        fg       fga       fg%        3p       3pa       3p%  \
0        NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1        NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2        NaN       NaN       NaN       NaN       NaN       NaN       NaN   
3        NaN       NaN       NaN       NaN       NaN       NaN       NaN   
4        NaN       NaN       NaN       NaN       NaN       NaN       NaN   
...      ...       ...       ...       ...       ...       ...       ...   
17767  0.000  0.565909  0.432353  0.534928  0.437931  0.415152  0.470665   
17768  0.000  0.559091  0.411765  0.544258  0.424138  0.451515  0.431235   
17769  0.025  0.493182  0.489706  0.403110  0.503448  0.574242  0.413420   
17770  0.000  0.522727  0.426471  0.493062  0.382759  0.413636  0.415914   
17771  0.025  0.552273  0.402941  0.540670  0.324138  0.337879  0.419715   

             ft       fta       ft%  ...  stl%_max_opp  blk%_max_opp  \
0           NaN       NaN       NaN  ...           NaN           NaN   
1           NaN       NaN       NaN  ...           NaN           NaN   
2           NaN       NaN       NaN  ...           NaN           NaN   
3           NaN       NaN       NaN  ...           NaN           NaN   
4           NaN       NaN       NaN  ...           NaN           NaN   
...         ...       ...       ...  ...           ...           ...   
17767  0.381395  0.339683  0.741190  ...        0.0509        0.0803   
17768  0.325581  0.314286  0.678530  ...        0.0721        0.0583   
17769  0.272093  0.250794  0.715986  ...        0.0627        0.0942   
17770  0.381395  0.338095  0.738390  ...        0.0610        0.0593   
17771  0.434884  0.363492  0.786114  ...        0.0387        0.0820   

       tov%_max_opp  usg%_max_opp  ortg_max_opp  drtg_max_opp  total_opp  \
0               NaN           NaN           NaN           NaN        NaN   
1               NaN           NaN           NaN           NaN        NaN   
2               NaN           NaN           NaN           NaN        NaN   
3               NaN           NaN           NaN           NaN        NaN   
4               NaN           NaN           NaN           NaN        NaN   
...             ...           ...           ...           ...        ...   
17767      0.220021      0.159564      0.416114      0.591765   0.500000   
17768      0.278302      0.184082      0.446445      0.552941   0.438462   
17769      0.395597      0.191528      0.490047      0.515294   0.515385   
17770      0.622956      0.147368      0.374408      0.542353   0.418269   
17771      0.395807      0.181130      0.522275      0.554118   0.448077   

       home_opp  won  season  
0           NaN  NaN     NaN  
1           NaN  NaN     NaN  
2           NaN  NaN     NaN  
3           NaN  NaN     NaN  
4           NaN  NaN     NaN  
...         ...  ...     ...  
17767       0.6  0.6  2022.0  
17768       0.6  0.6  2022.0  
17769       0.5  0.2  2022.0  
17770       0.4  0.8  2022.0  
17771       0.4  0.7  2022.0  

[17772 rows x 138 columns]

Combine rolling columns with regular columns

In [46]:
#renaming the rolling columns so it doesnt overlap with the existing columns 
#_x indicates its the rolling average columns for the previous x games
rolling_cols = [f"{col}_x" for col in df_rolling.columns]
#rename columns by assigning to the .columns property 
df_rolling.columns = rolling_cols

#Concatonate the original data frame and the data frame with the rolling averages
#axis=1 combines the data frames using the columns to combine them it will make a wide data frame 
#axis=0 makes a tall data frame by putting the rows together
dataFrame = pd.concat([dataFrame, df_rolling], axis=1)

In [47]:
dataFrame

mp        fg       fga       fg%        3p       3pa       3p%  \
0      0.0  0.318182  0.397059  0.263158  0.172414  0.181818  0.371734   
1      0.0  0.545455  0.470588  0.473684  0.275862  0.333333  0.365796   
2      0.0  0.386364  0.455882  0.303828  0.517241  0.439394  0.540380   
3      0.0  0.272727  0.352941  0.239234  0.206897  0.227273  0.375297   
4      0.0  0.340909  0.367647  0.313397  0.344828  0.424242  0.371734   
...    ...       ...       ...       ...       ...       ...       ...   
17767  0.0  0.545455  0.544118  0.416268  0.551724  0.515152  0.500000   
17768  0.0  0.636364  0.367647  0.679426  0.655172  0.560606  0.549881   
17769  0.0  0.386364  0.617647  0.200957  0.413793  0.575758  0.339667   
17770  0.0  0.500000  0.588235  0.337321  0.379310  0.469697  0.372922   
17771  0.0  0.363636  0.382353  0.330144  0.413793  0.348485  0.527316   

             ft       fta       ft%  ...  stl%_max_opp_x  blk%_max_opp_x  \
0      0.488372  0.444444  0.718786  ...             NaN             NaN   
1      0.186047  0.142857  0.883314  ...             NaN             NaN   
2      0.372093  0.269841  0.934656  ...             NaN             NaN   
3      0.279070  0.253968  0.725788  ...             NaN             NaN   
4      0.348837  0.269841  0.870478  ...             NaN             NaN   
...         ...       ...       ...  ...             ...             ...   
17767  0.418605  0.317460  0.889148  ...          0.0509          0.0803   
17768  0.581395  0.460317  0.844807  ...          0.0721          0.0583   
17769  0.488372  0.412698  0.784131  ...          0.0627          0.0942   
17770  0.325581  0.238095  0.927655  ...          0.0610          0.0593   
17771  0.581395  0.460317  0.844807  ...          0.0387          0.0820   

       tov%_max_opp_x  usg%_max_opp_x  ortg_max_opp_x  drtg_max_opp_x  \
0                 NaN             NaN             NaN             NaN   
1                 NaN             NaN             NaN             NaN   
2                 NaN             NaN             NaN             NaN   
3                 NaN             NaN             NaN             NaN   
4                 NaN             NaN             NaN             NaN   
...               ...             ...             ...             ...   
17767        0.220021        0.159564        0.416114        0.591765   
17768        0.278302        0.184082        0.446445        0.552941   
17769        0.395597        0.191528        0.490047        0.515294   
17770        0.622956        0.147368        0.374408        0.542353   
17771        0.395807        0.181130        0.522275        0.554118   

       total_opp_x  home_opp_x  won_x  season_x  
0              NaN         NaN    NaN       NaN  
1              NaN         NaN    NaN       NaN  
2              NaN         NaN    NaN       NaN  
3              NaN         NaN    NaN       NaN  
4              NaN         NaN    NaN       NaN  
...            ...         ...    ...       ...  
17767     0.500000         0.6    0.6    2022.0  
17768     0.438462         0.6    0.6    2022.0  
17769     0.515385         0.5    0.2    2022.0  
17770     0.418269         0.4    0.8    2022.0  
17771     0.448077         0.4    0.7    2022.0  

[17772 rows x 280 columns]

In [48]:
#this drops any rows that have missing values in them
#this will drop the first x amount of games in the season
dataFrame = dataFrame.dropna()

In [49]:
dataFrame
#no more missing values

mp        fg       fga       fg%        3p       3pa       3p%  \
1277   0.00  0.522727  0.470588  0.449761  0.448276  0.409091  0.497625   
1401   0.25  0.568182  0.558824  0.430622  0.551724  0.590909  0.441805   
1467   0.00  0.590909  0.441176  0.552632  0.482759  0.515152  0.437055   
1492   0.00  0.386364  0.308824  0.418660  0.275862  0.257576  0.452494   
1495   0.00  0.477273  0.338235  0.509569  0.310345  0.196970  0.628266   
...     ...       ...       ...       ...       ...       ...       ...   
17767  0.00  0.545455  0.544118  0.416268  0.551724  0.515152  0.500000   
17768  0.00  0.636364  0.367647  0.679426  0.655172  0.560606  0.549881   
17769  0.00  0.386364  0.617647  0.200957  0.413793  0.575758  0.339667   
17770  0.00  0.500000  0.588235  0.337321  0.379310  0.469697  0.372922   
17771  0.00  0.363636  0.382353  0.330144  0.413793  0.348485  0.527316   

             ft       fta       ft%  ...  stl%_max_opp_x  blk%_max_opp_x  \
1277   0.348837  0.285714  0.815636  ...          0.0589          0.0913   
1401   0.372093  0.380952  0.626604  ...          0.0801          0.1957   
1467   0.255814  0.253968  0.656943  ...          0.0679          0.0712   
1492   0.302326  0.222222  0.921820  ...          0.0635          0.0834   
1495   0.418605  0.380952  0.719953  ...          0.0698          0.0494   
...         ...       ...       ...  ...             ...             ...   
17767  0.418605  0.317460  0.889148  ...          0.0509          0.0803   
17768  0.581395  0.460317  0.844807  ...          0.0721          0.0583   
17769  0.488372  0.412698  0.784131  ...          0.0627          0.0942   
17770  0.325581  0.238095  0.927655  ...          0.0610          0.0593   
17771  0.581395  0.460317  0.844807  ...          0.0387          0.0820   

       tov%_max_opp_x  usg%_max_opp_x  ortg_max_opp_x  drtg_max_opp_x  \
1277         0.388574        0.196534        0.391469        0.527059   
1401         0.332914        0.145956        0.363507        0.561176   
1467         0.348218        0.127599        0.319905        0.530588   
1492         0.373899        0.138126        0.382464        0.454118   
1495         0.372537        0.153659        0.293365        0.488235   
...               ...             ...             ...             ...   
17767        0.220021        0.159564        0.416114        0.591765   
17768        0.278302        0.184082        0.446445        0.552941   
17769        0.395597        0.191528        0.490047        0.515294   
17770        0.622956        0.147368        0.374408        0.542353   
17771        0.395807        0.181130        0.522275        0.554118   

       total_opp_x  home_opp_x  won_x  season_x  
1277      0.313462         0.5    0.7    2016.0  
1401      0.445192         0.7    0.8    2017.0  
1467      0.386538         0.6    0.6    2021.0  
1492      0.243269         0.3    0.9    2016.0  
1495      0.291346         0.5    0.8    2016.0  
...            ...         ...    ...       ...  
17767     0.500000         0.6    0.6    2022.0  
17768     0.438462         0.6    0.6    2022.0  
17769     0.515385         0.5    0.2    2022.0  
17770     0.418269         0.4    0.8    2022.0  
17771     0.448077         0.4    0.7    2022.0  

[15882 rows x 280 columns]

Improving accuracy- giving the algorithm some information we know in advance about the next game
In the real world we know if the next NBA basketball game will be at home or away and we know who the opponent will be
I will give the algorithm this information so it can make better judgements

When we call add_col it will run group by team because I want to get the next game home value for the specific team. Then it will apply the shift_col function and it will pass in the team data frame so data just for one team, finally i pass in the column name. 

The shift_col takes that column and shifts it one position by taking the next rows value and shift it back to the current row.

For example it will look at the home column find if the next game will be played home or away and put that in the current row.

This is also done for the columns team_next_opponent and date_next.

This is only done for information known in advance. For future games I know the team they will play against the date that they are going to play and if the team is home or away. I don't know for example the specific performance stats of the player or the team.

In [50]:
#this function shift_colum takes in a team data frame and a column name
def shift_col(team, col_name):
    #creates a next column by taking the value from the next game and shift it back one row 
    next_col = team[col_name].shift(-1)
    return next_col

#function to add a column based on the shift_col function
#this will take in the main dataframe and the column name to aggregate based on 
def add_col(dataFrame, col_name):
    #group by teams 
    #applying the shift column function
    return dataFrame.groupby("team", group_keys=False).apply(lambda x: shift_col(x,col_name))

#using add_col function to give the algorithm information about the next game
#this is the information that everyone knows in advance about the games

#this column will indicate if the team will be home or away in its next game
#this creates a new column called home_next
#the output of this add column function will be assigned to the "home_next"
#passing in the data frame and the column "home"
dataFrame["home_next"] = add_col(dataFrame, "home")
#next opponent
dataFrame["team_opp_next"] = add_col(dataFrame, "team_opp")
#next date that they are going to play
dataFrame["date_next"] = add_col(dataFrame, "date")

#the copy warning can be a result of taking too many slices of a data frame and then you try to assign to it
#sometimes things will fail sometimes things will not

C:\Users\Garnet\AppData\Local\Temp\ipykernel_20912\3972476026.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataFrame["home_next"] = add_col(dataFrame, "home")
C:\Users\Garnet\AppData\Local\Temp\ipykernel_20912\3972476026.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataFrame["team_opp_next"] = add_col(dataFrame, "team_opp")
C:\Users\Garnet\AppData\Local\Temp\ipykernel_20912\3972476026.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using

In [51]:
dataFrame

mp        fg       fga       fg%        3p       3pa       3p%  \
1277   0.00  0.522727  0.470588  0.449761  0.448276  0.409091  0.497625   
1401   0.25  0.568182  0.558824  0.430622  0.551724  0.590909  0.441805   
1467   0.00  0.590909  0.441176  0.552632  0.482759  0.515152  0.437055   
1492   0.00  0.386364  0.308824  0.418660  0.275862  0.257576  0.452494   
1495   0.00  0.477273  0.338235  0.509569  0.310345  0.196970  0.628266   
...     ...       ...       ...       ...       ...       ...       ...   
17767  0.00  0.545455  0.544118  0.416268  0.551724  0.515152  0.500000   
17768  0.00  0.636364  0.367647  0.679426  0.655172  0.560606  0.549881   
17769  0.00  0.386364  0.617647  0.200957  0.413793  0.575758  0.339667   
17770  0.00  0.500000  0.588235  0.337321  0.379310  0.469697  0.372922   
17771  0.00  0.363636  0.382353  0.330144  0.413793  0.348485  0.527316   

             ft       fta       ft%  ...  usg%_max_opp_x  ortg_max_opp_x  \
1277   0.348837  0.285714  0.815636  ...        0.196534        0.391469   
1401   0.372093  0.380952  0.626604  ...        0.145956        0.363507   
1467   0.255814  0.253968  0.656943  ...        0.127599        0.319905   
1492   0.302326  0.222222  0.921820  ...        0.138126        0.382464   
1495   0.418605  0.380952  0.719953  ...        0.153659        0.293365   
...         ...       ...       ...  ...             ...             ...   
17767  0.418605  0.317460  0.889148  ...        0.159564        0.416114   
17768  0.581395  0.460317  0.844807  ...        0.184082        0.446445   
17769  0.488372  0.412698  0.784131  ...        0.191528        0.490047   
17770  0.325581  0.238095  0.927655  ...        0.147368        0.374408   
17771  0.581395  0.460317  0.844807  ...        0.181130        0.522275   

       drtg_max_opp_x  total_opp_x  home_opp_x  won_x  season_x  home_next  \
1277         0.527059     0.313462         0.5    0.7    2016.0        0.0   
1401         0.561176     0.445192         0.7    0.8    2017.0        0.0   
1467         0.530588     0.386538         0.6    0.6    2021.0        0.0   
1492         0.454118     0.243269         0.3    0.9    2016.0        1.0   
1495         0.488235     0.291346         0.5    0.8    2016.0        0.0   
...               ...          ...         ...    ...       ...        ...   
17767        0.591765     0.500000         0.6    0.6    2022.0        NaN   
17768        0.552941     0.438462         0.6    0.6    2022.0        NaN   
17769        0.515294     0.515385         0.5    0.2    2022.0        NaN   
17770        0.542353     0.418269         0.4    0.8    2022.0        NaN   
17771        0.554118     0.448077         0.4    0.7    2022.0        NaN   

       team_opp_next   date_next  
1277             WAS  03/03/2017  
1401             SAS  03/05/2017  
1467             ORL  03/03/2021  
1492             HOU  03/05/2017  
1495             BRK  03/03/2022  
...              ...         ...  
17767            NaN         NaN  
17768            NaN         NaN  
17769            NaN         NaN  
17770            NaN         NaN  
17771            NaN         NaN  

[15882 rows x 283 columns]

In [52]:
#this makes sure that the data frame is no longer pointing to a slice of itself
dataFrame= dataFrame.copy()

In [53]:
dataFrame

mp        fg       fga       fg%        3p       3pa       3p%  \
1277   0.00  0.522727  0.470588  0.449761  0.448276  0.409091  0.497625   
1401   0.25  0.568182  0.558824  0.430622  0.551724  0.590909  0.441805   
1467   0.00  0.590909  0.441176  0.552632  0.482759  0.515152  0.437055   
1492   0.00  0.386364  0.308824  0.418660  0.275862  0.257576  0.452494   
1495   0.00  0.477273  0.338235  0.509569  0.310345  0.196970  0.628266   
...     ...       ...       ...       ...       ...       ...       ...   
17767  0.00  0.545455  0.544118  0.416268  0.551724  0.515152  0.500000   
17768  0.00  0.636364  0.367647  0.679426  0.655172  0.560606  0.549881   
17769  0.00  0.386364  0.617647  0.200957  0.413793  0.575758  0.339667   
17770  0.00  0.500000  0.588235  0.337321  0.379310  0.469697  0.372922   
17771  0.00  0.363636  0.382353  0.330144  0.413793  0.348485  0.527316   

             ft       fta       ft%  ...  usg%_max_opp_x  ortg_max_opp_x  \
1277   0.348837  0.285714  0.815636  ...        0.196534        0.391469   
1401   0.372093  0.380952  0.626604  ...        0.145956        0.363507   
1467   0.255814  0.253968  0.656943  ...        0.127599        0.319905   
1492   0.302326  0.222222  0.921820  ...        0.138126        0.382464   
1495   0.418605  0.380952  0.719953  ...        0.153659        0.293365   
...         ...       ...       ...  ...             ...             ...   
17767  0.418605  0.317460  0.889148  ...        0.159564        0.416114   
17768  0.581395  0.460317  0.844807  ...        0.184082        0.446445   
17769  0.488372  0.412698  0.784131  ...        0.191528        0.490047   
17770  0.325581  0.238095  0.927655  ...        0.147368        0.374408   
17771  0.581395  0.460317  0.844807  ...        0.181130        0.522275   

       drtg_max_opp_x  total_opp_x  home_opp_x  won_x  season_x  home_next  \
1277         0.527059     0.313462         0.5    0.7    2016.0        0.0   
1401         0.561176     0.445192         0.7    0.8    2017.0        0.0   
1467         0.530588     0.386538         0.6    0.6    2021.0        0.0   
1492         0.454118     0.243269         0.3    0.9    2016.0        1.0   
1495         0.488235     0.291346         0.5    0.8    2016.0        0.0   
...               ...          ...         ...    ...       ...        ...   
17767        0.591765     0.500000         0.6    0.6    2022.0        NaN   
17768        0.552941     0.438462         0.6    0.6    2022.0        NaN   
17769        0.515294     0.515385         0.5    0.2    2022.0        NaN   
17770        0.542353     0.418269         0.4    0.8    2022.0        NaN   
17771        0.554118     0.448077         0.4    0.7    2022.0        NaN   

       team_opp_next   date_next  
1277             WAS  03/03/2017  
1401             SAS  03/05/2017  
1467             ORL  03/03/2021  
1492             HOU  03/05/2017  
1495             BRK  03/03/2022  
...              ...         ...  
17767            NaN         NaN  
17768            NaN         NaN  
17769            NaN         NaN  
17770            NaN         NaN  
17771            NaN         NaN  

[15882 rows x 283 columns]

I now have information on how the home team has done in the last x games. Now I'm going to pull in the same information for the opponent team.

This merger is saying: there is a team that I'm trying to predict data for in the left_on, I will find their next opponent and merge that opponents information together with the current home teams information. I want to merge based on the next games because the previous x amount of games for the opponent team

In [54]:
#to pull information from the opposing team also I use the merge commanand
#this will do an inner join,on the columns we are trying to merge
#the left side is going to be the team I am trying to predict
#the right side is going to be the team that is playings' next opponent
#pulling in the rolling columns and information about the next game to merge based on that  
full = dataFrame.merge(
    dataFrame[rolling_cols + ["team_opp_next", "date_next", "team"]], 
#merging based on the team column and the date next column
    left_on=["team", "date_next"], 
#merging based on the team_opp_next column and the date_next column 
    right_on=["team_opp_next", "date_next"]
)

x is the amount of previous games the model is looking at 
In this example the first _x is equal to the previuos 10 games
In cases where two columns are named the same the left side of the merge will get the underscore x the right side will get the underscore y

In [55]:
full
#the row count went down because the empty values where those were the last games of that season were removed

mp        fg       fga       fg%        3p       3pa       3p%  \
0      0.25  0.568182  0.558824  0.430622  0.551724  0.590909  0.441805   
1      0.00  0.386364  0.308824  0.418660  0.275862  0.257576  0.452494   
2      0.00  0.431818  0.367647  0.425837  0.724138  0.621212  0.554632   
3      0.00  0.454545  0.205882  0.617225  0.241379  0.333333  0.319477   
4      0.00  0.318182  0.264706  0.368421  0.275862  0.348485  0.351544   
...     ...       ...       ...       ...       ...       ...       ...   
15715  0.00  0.363636  0.352941  0.354067  0.413793  0.454545  0.419240   
15716  0.00  0.409091  0.323529  0.435407  0.413793  0.287879  0.619952   
15717  0.00  0.500000  0.352941  0.523923  0.586207  0.439394  0.611639   
15718  0.00  0.590909  0.382353  0.607656  0.448276  0.484848  0.428741   
15719  0.00  0.590909  0.529412  0.478469  0.379310  0.439394  0.395487   

             ft       fta       ft%  ...  tov%_max_opp_x_y  usg%_max_opp_x_y  \
0      0.372093  0.380952  0.626604  ...          0.373899          0.138126   
1      0.302326  0.222222  0.921820  ...          0.332914          0.145956   
2      0.395349  0.285714  0.938156  ...          0.316562          0.211810   
3      0.372093  0.333333  0.735123  ...          0.313312          0.107831   
4      0.255814  0.285714  0.570595  ...          0.352725          0.174711   
...         ...       ...       ...  ...               ...               ...   
15715  0.279070  0.285714  0.631272  ...          0.470755          0.120154   
15716  0.279070  0.365079  0.465578  ...          0.355870          0.153017   
15717  0.139535  0.142857  0.649942  ...          0.285430          0.201540   
15718  0.418605  0.380952  0.719953  ...          0.422013          0.109628   
15719  0.395349  0.365079  0.708285  ...          0.294969          0.172529   

       ortg_max_opp_x_y  drtg_max_opp_x_y  total_opp_x_y  home_opp_x_y  \
0              0.382464          0.454118       0.243269           0.3   
1              0.363507          0.561176       0.445192           0.7   
2              0.423223          0.483529       0.418269           0.7   
3              0.410427          0.440000       0.365385           0.4   
4              0.498104          0.535294       0.446154           0.5   
...                 ...               ...            ...           ...   
15715          0.387678          0.548235       0.491346           0.3   
15716          0.415640          0.428235       0.550962           0.7   
15717          0.358768          0.496471       0.411538           0.5   
15718          0.482464          0.523529       0.440385           0.4   
15719          0.443602          0.545882       0.543269           0.7   

       won_x_y  season_x_y  team_opp_next_y  team_y  
0          0.9      2016.0              HOU     SAS  
1          0.8      2017.0              SAS     HOU  
2          0.5      2016.0              ATL     POR  
3          0.4      2016.0              CHI     CHO  
4          0.6      2022.0              PHO     GSW  
...        ...         ...              ...     ...  
15715      0.6      2019.0              UTA     MIN  
15716      0.3      2019.0              CLE     ATL  
15717      0.6      2020.0              PHO     BOS  
15718      0.7      2020.0              HOU     MIA  
15719      0.3      2020.0              SAC     DAL  

[15720 rows x 423 columns]

In [56]:
#to show how the merge works
#meant to be 15774 rows x 5 columns
#dates are not in order
full[["team_x","team_opp_next_x", "team_y","team_opp_next_y", "date_next"]]

team_x team_opp_next_x team_y team_opp_next_y   date_next
0        HOU             SAS    SAS             HOU  03/05/2017
1        SAS             HOU    HOU             SAS  03/05/2017
2        ATL             POR    POR             ATL  03/05/2021
3        CHI             CHO    CHO             CHI  03/11/2015
4        PHO             GSW    GSW             PHO  03/12/2016
...      ...             ...    ...             ...         ...
15715    UTA             MIN    MIN             UTA  31/12/2021
15716    CLE             ATL    ATL             CLE  31/12/2021
15717    PHO             BOS    BOS             PHO  31/12/2021
15718    HOU             MIA    MIA             HOU  31/12/2021
15719    SAC             DAL    DAL             SAC  31/12/2021

[15720 rows x 5 columns]

Now I will run through the sequential feature selector again to find the 30 prediictors from the new set. Seeing which 30 columns out of the 423 columns is best for the model 

In [57]:
#the set of columns I dont want to pass into the model
#object dates are columns that arent numeric like the team codes or dates 
#removed columns includes the target which I dont want to include into the model when training it 
removed_columns = list(full.columns[full.dtypes == "object"]) + removed_columns

In [58]:
#these columns either wont work in my model or would invalidate the prediction of the target
removed_columns

['team_x',
 'team_opp',
 'date',
 'team_opp_next_x',
 'date_next',
 'team_opp_next_y',
 'team_y',
 'season',
 'date',
 'won',
 'target',
 'team',
 'team_opp']

In [59]:
#taking all of the columns from the full data frame except for the ones that are in the removed columns list
#~negation operation
#is this column in removed columns if its not then select it into the selected columns list
selected_columns = full.columns[~full.columns.isin(removed_columns)]

In [ ]:
#selected_columns are the predictors that are passed in to try and predict the target
sfs.fit(full[selected_columns], full["target"])
#the sequential feature selector is going to use the best 30 features that will give the most value for predicting our target

In [ ]:
#returns a boolean that indicates whether that feature was selected or not
#index selected columns list by this index
predictors = list(selected_columns[sfs.get_support()])

In [ ]:
predictors
#there are a few predictors from both sides the home and the oppponents side

In [ ]:
#generate predictions
#passses in full data frame, ridge regression model and predictors 
predictions = backtest(full, rc, predictors)

In [ ]:
#define accuracy score
accuracy_score(predictions["actual"], predictions["prediction"])
#now the accuracy has met the baseline of 57%

In [ ]:
from sklearn.metrics import precision_score 
precision_score(actual["target"], predictors)

In [ ]:
precision_score(actual["target"], predictors)

In [ ]:
#to improve accuracy
#dates are not in order go back and fix this 
#try using a more powerful model than ridge regression to improve the accuracy such as "xG-boost" or decision tree or random forrest classifier
#try out different amounts of features that are selected
#try backwards feature selection results in a better set of features
#try different ratios for example looking at the last 5 or 20 games instead of the last 10 and compare the differences
#to predict future values
#rerun the webscraping code that produced the excel document to get the most up to date box scores
# I will need to manually fill in the home_next, team_opp_next, date_next for the null/empty values
#rerun full = dataFrame.merge to include the rows: home_next, team_opp_next, date_next after the merge with no nulls
#rerun sfs.fit(full[selected_columns], full["target"]) predictions and look at any rows where the target is 2(for future games) we can find future predictions based on that